In [1]:
import pandas as pd
import numpy as np 
import plotly.graph_objects as go 
import plotly.express as px 
from scipy.stats import stats

In [2]:
ab_nyc = pd.read_csv(r"C:\potfolio\Aribnb_price_prediction\data\raw\AB_NYC_2019.csv", sep = ",", index_col=False)
ab_nyc.head()

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
0,2539,Clean & quiet apt home by the park,2787,John,Brooklyn,Kensington,40.64749,-73.97237,Private room,149,1,9,2018-10-19,0.21,6,365
1,2595,Skylit Midtown Castle,2845,Jennifer,Manhattan,Midtown,40.75362,-73.98377,Entire home/apt,225,1,45,2019-05-21,0.38,2,355
2,3647,THE VILLAGE OF HARLEM....NEW YORK !,4632,Elisabeth,Manhattan,Harlem,40.80902,-73.94190,Private room,150,3,0,NaN,NaN,1,365
3,3831,Cozy Entire Floor of Brownstone,4869,LisaRoxanne,Brooklyn,Clinton Hill,40.68514,-73.95976,Entire home/apt,89,1,270,2019-07-05,4.64,1,194
4,5022,Entire Apt: Spacious Studio/Loft by central park,7192,Laura,Manhattan,East Harlem,40.79851,-73.94399,Entire home/apt,80,10,9,2018-11-19,0.10,1,0


In [3]:
ab_nyc = ab_nyc[ ab_nyc["price"] > 0].copy()

In [4]:
ab_nyc["log_price"] = np.log(ab_nyc["price"])

In [5]:
print(ab_nyc.info())

<class 'pandas.core.frame.DataFrame'>
Index: 48884 entries, 0 to 48894
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48884 non-null  int64  
 1   name                            48868 non-null  object 
 2   host_id                         48884 non-null  int64  
 3   host_name                       48863 non-null  object 
 4   neighbourhood_group             48884 non-null  object 
 5   neighbourhood                   48884 non-null  object 
 6   latitude                        48884 non-null  float64
 7   longitude                       48884 non-null  float64
 8   room_type                       48884 non-null  object 
 9   price                           48884 non-null  int64  
 10  minimum_nights                  48884 non-null  int64  
 11  number_of_reviews               48884 non-null  int64  
 12  last_review                     38833

In [6]:
ab_nyc.describe()

,id,host_id,latitude,longitude,price,minimum_nights,number_of_reviews,reviews_per_month,calculated_host_listings_count,availability_365,log_price
count,4.888400e+04,4.888400e+04,48884.000000,48884.000000,48884.000000,48884.000000,48884.000000,38833.000000,48884.000000,48884.000000,48884.000000
mean,1.901679e+07,6.762203e+07,40.728953,-73.952176,152.755053,7.029887,23.271991,1.373128,7.144628,112.779498,4.727031
std,1.098432e+07,7.861666e+07,0.054532,0.046159,240.170260,20.512224,44.551331,1.680391,32.956185,131.627271,0.698225
min,2.539000e+03,2.438000e+03,40.499790,-74.244420,10.000000,1.000000,0.000000,0.010000,1.000000,0.000000,2.302585
25%,9.470548e+06,7.817310e+06,40.690100,-73.983080,69.000000,1.000000,1.000000,0.190000,1.000000,0.000000,4.234107
50%,1.967574e+07,3.079257e+07,40.723080,-73.955685,106.000000,3.000000,5.000000,0.720000,1.000000,45.000000,4.663439
75%,2.915297e+07,1.074344e+08,40.763120,-73.936290,175.000000,5.000000,24.000000,2.020000,2.000000,227.000000,5.164786
max,3.648724e+07,2.743213e+08,40.913060,-73.712990,10000.000000,1250.000000,629.000000,58.500000,327.000000,365.000000,9.210340


In [7]:
print(ab_nyc[["price","log_price"]].describe())

              price     log_price
count  48884.000000  48884.000000
mean     152.755053      4.727031
std      240.170260      0.698225
min       10.000000      2.302585
25%       69.000000      4.234107
50%      106.000000      4.663439
75%      175.000000      5.164786
max    10000.000000      9.210340


In [8]:
fig = px.histogram(ab_nyc, x="price", title="Price distribution", nbins=25)
fig.update_traces(marker = dict(color = "skyblue", line = dict(color = "black", width = 1)))
fig.show()

In [9]:
fig = px.histogram(ab_nyc, x="log_price", title="Log Price distribution", nbins=25)
fig.update_traces(marker = dict(color = "blue", line = dict(color = "white", width = 1)))
fig.show()

In [10]:
ab_nyc[ab_nyc["number_of_reviews"]==0][1:5]

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365,log_price
19,7750,Huge 2 BR Upper East Cental Park,17985,Sing,Manhattan,East Harlem,40.79685,-73.94872,Entire home/apt,190,7,0,NaN,NaN,2,249,5.247024
26,8700,Magnifique Suite au N de Manhattan - vue Cloitres,26394,Claude & Sophie,Manhattan,Inwood,40.86754,-73.92639,Private room,80,4,0,NaN,NaN,1,0,4.382027
36,11452,Clean and Quiet in Brooklyn,7355,Vt,Brooklyn,Bedford-Stuyvesant,40.68876,-73.94312,Private room,35,60,0,NaN,NaN,1,365,3.555348
38,11943,Country space in the city,45445,Harriet,Brooklyn,Flatbush,40.63702,-73.96327,Private room,150,1,0,NaN,NaN,1,365,5.010635


**Where ever the number of reviews is = 0 we have get last_review and review_per_month as NAN hence this is an MAR**

In [11]:
high_mean_value = ab_nyc.groupby("neighbourhood").agg({
    "price" : [ "median", "mean", "sum"],
    "log_price" : ["median", "mean", "sum"]
}).sort_values(by = ("price","median"), ascending=False)
high_mean_value.head(15)

price                     log_price                       
                   median        mean     sum    median      mean          sum
neighbourhood                                                                 
Fort Wadsworth      800.0  800.000000     800  6.684612  6.684612     6.684612
Woodrow             700.0  700.000000     700  6.551080  6.551080     6.551080
Tribeca             295.0  490.638418   86843  5.686975  5.754987  1018.632647
Neponsit            274.0  274.666667     824  5.613128  5.589793    16.769379
NoHo                250.0  295.717949   23066  5.521461  5.530059   431.344607
Willowbrook         249.0  249.000000     249  5.517453  5.517453     5.517453
Flatiron District   225.0  341.925000   27354  5.416100  5.536889   442.951112
Midtown             210.0  282.719094  436801  5.347108  5.416535  8368.546650
Financial District  200.0  225.490591  167765  5.298317  5.270275  3921.084353
West Village        200.0  267.682292  205580  5.298317  5.383147  4134.256958
Chelsea             199.0  249.738544  277959  5.293305  5.273853  5869.798176
SoHo                199.0  287.103352  102783  5.293305  5.346598  1914.082185
Greenwich Village   197.5  263.405612  103255  5.285735  5.282923  2070.905768
Battery Park City   195.0  367.557143   25729  5.273000  5.266488   368.654156
Breezy Point        195.0  213.333333     640  5.273000  5.355820    16.067460

In [12]:
ab_nyc.groupby("neighbourhood_group")["price"].median().sort_values(ascending=False)


neighbourhood_group
Manhattan        150.0
Brooklyn          90.0
Queens            75.0
Staten Island     75.0
Bronx             65.0
Name: price, dtype: float64

In [13]:
group_room_price = ab_nyc.groupby("room_type")["price"].median().reset_index().sort_values(by = "price",ascending=False)
fig = px.box(ab_nyc, x="room_type", y ="price", title="Box plot for types of rooms",points="outliers")
fig.update_traces(marker = dict(color = "red", line =dict(color = "black", width = 1)))
fig.update_layout(height = 600)
fig.show()

In [14]:
fig = px.box(ab_nyc, x="room_type", y ="log_price", title="Box plot for types of rooms",points="outliers")
fig.update_traces(marker = dict(color = "red", line =dict(color = "black", width = 1)))
fig.update_layout(height = 600)
fig.show()

In [15]:
two_way = ab_nyc.groupby(["neighbourhood_group", "room_type"])["price"].median().reset_index().pivot(index="neighbourhood_group", columns="room_type", values="price")
two_way

room_type,Entire home/apt,Private room,Shared room
neighbourhood_group,,,
Bronx,100.0,54.0,40.0
Brooklyn,145.0,65.0,36.0
Manhattan,191.0,90.0,69.0
Queens,120.0,60.0,37.0
Staten Island,100.0,50.0,30.0


In [16]:
fig = px.imshow(
    two_way,
    text_auto=True,
    color_continuous_scale="RdBu_r",
    title="Median Price by Room Type and Neighbourhood Group"
)
fig.update_layout(height=500, width=1000)
fig.show()


**1. Manhattan is the most expensive**

    → For all room types
    → Entire home/apt median ≈ 200+
    → Private rooms also expensive (~100)

**2. Brooklyn is second**

    → Entire home/apt ≈ 150
    → Private rooms ≈ 75

**3. Queens, Bronx, Staten Island are cheaper**

    Queens has mid prices

    Bronx lowest overall

    Staten Island cheap & low supply

**4. Room Type effect is consistent**

    Across all areas:

    Entire home/apt > Private room > Shared room

    This shows that room type has global effect, while neighbourhood has local variation.

In [17]:
df = ab_nyc.dropna(subset=["log_price"])

In [18]:
import statsmodels.api as sm 
from statsmodels.formula.api import ols 

model = ols(
    "log_price ~ C(neighbourhood_group) * C(room_type)", data = df
).fit()

anova_table = sm.stats.anova_lm(model , typ =2)
anova_table

,sum_sq,df,F,PR(>F)
C(neighbourhood_group),1633.625851,4.0,1538.039709,0.000000e+00
C(room_type),7681.851332,2.0,14464.747091,0.000000e+00
C(neighbourhood_group):C(room_type),28.829437,8.0,13.571290,7.901781e-20
Residual,12976.528051,48869.0,NaN,NaN
